src/metaswitch/clearwater/config_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,144 @@
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version, along with the "Special Exception" for use of
+# the program along with SSL, set forth below. This program is distributed
+# in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+import etcd
+from threading import Thread
+from time import sleep
+
+import urllib3
+import logging
+
+_log = logging.getLogger("etcd_sync")
+
+
+class EtcdSynchronizer(object):
+    PAUSE_BEFORE_RETRY = 30
+
+    def __init__(self, plugin, ip, alarm, etcd_ip=None, force_leave=False):
+        self._ip = ip
+        if etcd_ip:
+            self._client = etcd.Client(etcd_ip, 4000)
+        else:
+            self._client = etcd.Client(ip, 4000)
+        self._key = plugin.key()
+        self._plugin = plugin
+        self._alarm = alarm
+        self._index = None
+        self._terminate_flag = False
+        self.thread = Thread(target=self.main)
+        self.force_leave = force_leave
+
+    def main(self):
+        # Continue looping while the FSM is running.
+        while not self._terminate_flag:
+            # This blocks on changes to the watched key in etcd.
+            _log.debug("Waiting for change from etcd for key{}".format(
+                         self._key))
+            value = self.read_from_etcd()
+            if self._terminate_flag:
+                break
+
+            if len(value) > 0:
+                _log.debug("Got new config value from etcd:\n{}".format(value))
+                self._plugin.on_config_changed(value)
+                self._alarm.update_file(self._plugin.file())
```

Comment:
I'm not sure if its better to do this here or from within the plugin. If you do it here you're effectively saying that `on_config_changed` must not fail unless it raises an exception. Which is probably fine but the ABC needs a comment to that effect.